File: backups/sora-editor/editor/src/main/java/io/github/abc15018045126/sora/graphics/GraphicsCompat.java

package io.github.abc15018045126.sora.graphics;

import android.annotation.SuppressLint;
import android.graphics.Canvas;
import android.os.Build;

import androidx.annotation.NonNull;

public class GraphicsCompat {

    /**
     * {@link Canvas#drawTextRun(char[], int, int, int, int, float, float, boolean, android.graphics.Paint)} is also available on API 21 & 22,
     * but with hidden access.
     * As there is no hidden list checks in those API platforms, it's safe here to call the "New API".
     */
    @SuppressLint("NewApi")
    public static void drawTextRun(Canvas canvas, @NonNull char[] text, int index, int count, int contextIndex,
                                   int contextCount, float x, float y, boolean isRtl, @NonNull android.graphics.Paint paint) {
        canvas.drawTextRun(text, index, count, contextIndex, contextCount, x, y, isRtl, paint);
    }

    public static float getRunAdvance(Paint paint, char[] text, int start, int end, int contextStart, int contextEnd,
                                      boolean isRtl, int offset) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return paint.getRunAdvance(text, start, end, contextStart, contextEnd, isRtl, offset);
        } else {
            return paint.measureTextRunAdvance(text, start, offset, contextStart, contextEnd, isRtl);
        }
    }

}

File: android/sora-editor/editor/src/main/java/io/github/abc15018045126/sora/graphics/GraphicsCompat.kt
package io.github.abc15018045126.sora.graphics

import android.annotation.SuppressLint
import android.graphics.Canvas
import android.os.Build

object GraphicsCompat {


    @SuppressLint("NewApi")
    @JvmStatic
    fun drawTextRun(
        canvas: Canvas,
        text: CharArray,
        index: Int,
        count: Int,
        contextIndex: Int,
        contextCount: Int,
        x: Float,
        y: Float,
        isRtl: Boolean,
        paint: android.graphics.Paint
    ) {
        canvas.drawTextRun(text, index, count, contextIndex, contextCount, x, y, isRtl, paint)
    }

    @JvmStatic
    fun getRunAdvance(
        paint: Paint,
        text: CharArray,
        start: Int,
        end: Int,
        contextStart: Int,
        contextEnd: Int,
        isRtl: Boolean,
        offset: Int
    ): Float {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            paint.getRunAdvance(text, start, end, contextStart, contextEnd, isRtl, offset)
        } else {
            paint.measureTextRunAdvance(text, start, offset, contextStart, contextEnd, isRtl)
        }
    }
}
